### ISSUE TITLE
Clarify "cmd" value in tests descriptions

### ISSUE THREAD

**Issue by bkhl** [exercism/research_experiment_1#32]
https://github.com/exercism/research_experiment_1#config:

```json
{
  "tests": [
    {
      "name": "MyTestMethod",
      "cmd": "MyTests.MyTestMethod(1)"
    }
  ]
}
```

Could it be added to this document an explanation on what `cmd` should contain and what it's used for, we'll need it to figure what to put in here for the Bash exercises in https://github.com/exercism/bash/issues/412

**Comment by iHiD** [comment]
Hmmm, I'm not sure how to improve on [what's in the readme](https://github.com/exercism/research_experiment_1/blob/033cc0092d38f6ebd81d580e3cd4c87eca43dde7/README.md#test-messages). Any suggestions on how it could be clearer?

**Comment by bkhl** [comment]
The problem is that we can't tell what `MyTests.MyTestMethod(1)` is without knowing about the language the example is in.

Is that supposed to be some kind of command to run the test individually? Or just an identifier for the test (in which case the name is a bit misleading).

**Comment by iHiD** [comment]
So the README says:

> This means that when the code does not pass one of the tests, the user won't see the code of the test that failed or the input the code was called with. To fix this, we need to provide a bit of context when we show errors to the user. To achieve this, each exercise should define a tests key in its `.meta/config.json file`. This key is a mapping of the test method's name (the "name" value in the test runner output) to the command that we tell the user was run when the test fails.

So to rephrase, it is the entry point to a user's code that the test executes, and we show that to the user, who can then mentally link it the execution to the failure output they get. 

So, if we were to look at [TwoFer in Bash](https://github.com/exercism/bash/blob/master/exercises/two-fer/two_fer_test.sh#L28) as an example, we'd have something like this:

```json
[
  {
    "name": "no name given",
    "cmd": "run bash two_fer.sh"
  },
  {
    "name": "a name given",
    "cmd": "run bash two_fer.sh Alice"
  },
  {
    "name": "another name given",
    "cmd": "run bash two_fer.sh Bob"
  },
  {
    "name": "handle arg with spaces",
    "cmd": "run bash two_fer.sh \"John Smith\" \"Mary Ann\""
  }
]
```

Does that make any more sense?

**Comment by bkhl** [comment]
I think it does.

**Comment by iHiD** [comment]
Great. If there's a way to improve the clarity in the README, I'd be grateful for the PR :)

**Comment by ErikSchierboom** [comment]
@bkhl Am I correct in understanding that this issue can be closed?

**Comment by bkhl** [comment]
I think @iHiD implied he was waiting for a PR from me, see above.

I'm still slightly confused as the original examples look like they refer to test cases:

```json
{
  "tests": [
    {
      "name": "MyTestMethod",
      "cmd": "MyTests.MyTestMethod(1)"
    }
  ]
}
```

However, the wording and the suggestion from @iHiD  above about what to put in here for Bash seems to imply the messages should help the user re-run the *test* locally, rather than doing the equivalent manual test.

If we should give the user a method to rerun the *test*, rather than the code under test, we should be doing something more along these lines:

```json
{
  "tests": [
    {
      "name": "no name given",
      "cmd": "bats -f '^no name given$' test_two_fer.bats",
    },
    …
```

Consider my PR above a work in progress until we have cleared this out.

**Comment by guygastineau** [comment]
@bkhl that is terrifying! :rofl: 

Doing it that way would actually reduce the amount of parsing needed to automate this, but I would have a stroke if I saw that for my failing tests.  Granted I assume that is a regexp, but I am feeling a bit dramatic.

I guess it depends on how we define entry point.  The snippet above is an entry point into the bats tests, but those tests use a different entry point into the script we have student write.

So my first instinct is to say that @iHiD 's example is correct (incidentally also the way our PR already does it).  One issue with that is what to do with really long inputs for tests like `diamond`.  Personally, I don't think it is that big of an issue.  I expect most users just to rerun the whole test file (especially for a big input test).  I also see that we are covering our bases with the test name, so users should always be able to find the relevant section of the test file when working through failing tests.

If my opinion makes sense to everyone, then I am willing to let this be for now, and let us get moving forward ;)

**Comment by bkhl** [comment]
I just saw that the only way to pick a test with bats is by regexp. I didn't actually test the above, so not sure if you also have to escape spaces and whatnot.

**Comment by guygastineau** [comment]
It is awesome.  We all love when `regexp` help our lives work better.

I use them less than I used to though, so they are a bit jarring at first :laughing:

**Comment by guygastineau** [comment]
A more focused thought about how bash would interpret this on cases where the inputs are gigantic (I suppose this has some bearing for other languages too).

The way my PR for it is handling it now I just remove the input material when it is longer than one line.

Another approach could be to show the command for running the whole test.  I don't know which is preferable? Perhaps neither?

**Comment by ErikSchierboom** [comment]
Has this been closed by #47?

**Comment by ErikSchierboom** [comment]
I'm gonna close this issue as I think it has been fixed. If this is incorrect, please let me know!
